src/components/PlaybackControls.tsx

```diff
@@ -0,0 +1,249 @@
+  const { isPlaying, currentTime } = playbackState;
+
+  const nextFrameUnavailable = currentTime + timeStep >= lastFrameTime;
+  const previousFrameUnavailable = firstFrameTime + timeStep > currentTime;
+
+  const displayTimes = getDisplayTimes(playbackState, playbackData);
+  const timeInputWidth = getTimeInputWidth(displayTimes);
+
+  // useeffect to reset playback state when the last frame is reached
+  useEffect(() => {
+    controller.pause();
+    controller.gotoTime(firstFrameTime);
+    setPlaybackState({
+      currentTime: firstFrameTime,
+      isPlaying: false,
+    });
+  }, [nextFrameUnavailable]);
+
+  const skipToTime = (time: number) => {
+    const isTimePastLastFrame =
+      compareTimes(time, lastFrameTime, timeStep) === 1;
+    const isTimeBeforeFirstFrame =
+      compareTimes(time, firstFrameTime, timeStep) === -1;
+    if (isTimePastLastFrame || isTimeBeforeFirstFrame) {
+      return;
+    }
+    controller.gotoTime(time);
+  };
+
+  const playHandler = (timeOverride?: number) => {
+    const timeOverrideProvided = timeOverride !== undefined;
+    const newTime = timeOverrideProvided ? timeOverride : currentTime;
+    // this is the same as nextFrameUnavailable, but computing it here
+    // prevents a looping bug where playback gets stuck oscillating between
+    // first and last frame
+    const nextFrameIsPastLastFrame = newTime + timeStep >= lastFrameTime;
+    controller.pause();
+    if (nextFrameIsPastLastFrame) {
+      setPlaybackState({
+        isPlaying: false,
+        currentTime: firstFrameTime,
+      });
+      controller.gotoTime(firstFrameTime);
+    } else {
+      setPlaybackState({
+        isPlaying: true,
+        currentTime: newTime,
+      });
+      controller.playFromTime(newTime);
+      controller.resume();
+    }
+  };
+
+  const pauseHandler = () => {
+    controller.pause();
+    setPlaybackState({ ...playbackState, isPlaying: false });
+  };
+
+  const gotoNextFrame = () => {
+    controller.gotoTime(currentTime + timeStep);
+  };
+
+  const goToPreviousFrame = () => {
+    controller.gotoTime(currentTime - timeStep);
+  };
+
+  const handleTimeInputChange = (userInput: number | null): void => {
+    if (userInput !== null) {
+      setTimeInput(userInput as number);
+    }
+  };
+
+  const handleTimeInputKeyDown = (
+    event: React.KeyboardEvent<HTMLInputElement>
+  ): void => {
+    if (event.key === 'Enter' || event.key === 'Tab') {
+      if (typeof timeInput === 'number') {
+        setPlaybackState({
+          ...playbackState,
+          currentTime: timeInput / timeUnits.magnitude,
+        });
+        skipToTime(timeInput / timeUnits.magnitude);
+      }
+    }
+    if (event.key === 'Escape') {
+      const inputNumberComponent = event.target as HTMLElement;
+      inputNumberComponent.blur();
+    }
+  };
+
+  const handleSliderChange = (sliderValue: number | [number, number]): void => {
+    skipToTime(sliderValue as number);
```

Comment:
Should this do a check for the `[number, number]` case?

Reply:
Great catch, changing this here and in the website (where this code came from). Using `range={false}` prop on slider and arg can always be a number.